solidity/src/CadenceRandomConsumer.sol

```diff
@@ -0,0 +1,92 @@
+     * Emits a {RandomnessRequested} event.
+     *
+     * @return requestId The ID of the request.
+     */
+    function _requestRandomness() public returns (uint256) {
+        // Identify the request by the current request counter, incrementing first so implementations can use 0 for
+        // invalid requests - e.g. myRequests[msg.sender] == 0 means the caller has no pending requests
+        _requestCounter++;
+        uint256 requestId = _requestCounter;
+        // Store the heights at which the request was made. Note that the Flow block height and EVM block height are
+        // not the same. But since Flow blocks ultimately determine usage of secure randomness, we gate requests by
+        // Flow block height.
+        Request memory request = Request(_flowBlockHeight(), block.number);
+
+        // Store the request in the list of requests
+        _requests.push(request);
+
+        emit RandomnessRequested(requestId, request.flowHeight, request.evmHeight);
+
+        // Finally return the request ID
+        return requestId;
+    }
+
+    /**
+     * @dev This method serves as the reveal step in this contract's commit-reveal scheme
+     *
+     * Here a caller reveals a random number at least one block after the commit block
+     * This is because the random source for a Flow block is not available until after the finalization of that block.
+     * Note that the random source for a given Flow block is singular. In order to ensure that requests made at the same
+     * block height are unique, implementing contracts should use some pseudo-random method to generate a unique value
+     * from the seed along with a salt.
+     * Emits a {RandomnessFulfilled} event.
+     *
+     * @param requestId The ID of the randomness request to fulfill.
+     * @return randomResult The random value generated from the Flow block.
+     */
+    function _fulfillRandomness(uint32 requestId) internal returns (uint64) {
+        // Get the request details. Recall that request IDs are 1-indexed to allow for 0 to be used as an invalid value
+        uint32 requestIndex = requestId - 1;
+        require(requestIndex < _requests.length, "Invalid request ID - value exceeds the number of existing requests");
+        Request memory request = _requests[requestIndex];
+
+        // Ensure that the request is fulfilled at a Flow block height greater than the one at which the request was made
+        uint64 flowHeight = _flowBlockHeight();
+        require(request.flowHeight < flowHeight, "Cannot fulfill request until subsequent Flow network block height");
+
+        // Get the random source for the Flow block at which the request was made
+        uint64 randomResult = _getRandomSource(request.flowHeight); // returns bytes32
+        // Pack the randomResult into a uint64, hashing with the requestId to vary results across shared block heights
+        randomResult = uint64(uint256(keccak256(abi.encodePacked(randomResult, requestId))));
```

Comment:
That's where the PRG implementation detail can be added. We should add a comment to mention that the random source is only 32 bytes. Here only 8 bytes were required, but if the contract requires more random bytes, the source should be expanded into any arbitrary number of bytes using a PRG (which I believe is that we did on the Cadence example).

Reply:
Updated in [5861871](https://github.com/onflow/random-coin-toss/pull/20/commits/58618718823a0ee680050799e70ead633250c538)